storage/merkle/tree.go

```diff
@@ -185,6 +214,7 @@ PutLoop:
 		// if we have a leaf node, we reached a non-empty leaf
 		case *leaf:
 			n.val = append(make([]byte, 0, len(val)), val...)
+			n.cachedHashValue = nil
```

Comment:
To me, line 217 seems somewhat misleading and thereby inviting  bugs during future refactorings. Just in case we have cached values on any node along the path, these hashes would pertain to the old value and not the updated value. In my opinion, the root of the problem is that this implementation assumes nodes to be mutable. 

So during updates, the tree must have _no_ stored hashes, for which reason we also don't need to set the hash here to nil. By writing line 217, it creates the impression (at least for me) that a hash value could be cached already and it would suffice to just set the leaf's hash back to nil. But if there were _any_ hashes cached already, line 217 would just break the trie completely. 

```suggestion
			n.val = append(make([]byte, 0, len(val)), val...)
```

Reply:
this was not needed anymore, removed at [978fa92](https://github.com/onflow/flow-go/pull/3411/commits/978fa92be68bc5aa198e49a405b7e423946f55dc)